src/Parser/Druid/Feral/Modules/Features/SpellUsable.js

```diff
@@ -0,0 +1,121 @@
+
+const EARLY_BLEED_EXPIRE_TO_COUNT_AS_DEATH = 500;
+
+/**
+ * Predator:
+ * The cooldown on Tiger's Fury resets when a target dies with one of your Bleed effects active
+ * 
+ * We cannot directly detect enemy death, but can infer it from a bleed debuff ending early.
+ * Death is not the only cause of a bleed ending early, for instance a Monk's paralysis ability
+ * removes DoTs when it's applied to a target.
+ * We know for certain that an enemy died when Tiger's Fury is used earlier than should be possible.
+ * When that happens assume the most recent possible enemy death was when the cooldown reset.
+ */
+class SpellUsable extends CoreSpellUsable {
+  // BFA: replace this.combatants.selected with this.selectedCombatant, and remove combatants dependency
+  static dependencies = {
+    ...CoreSpellUsable.dependencies,
+    combatants: Combatants,
+  };
+
+  static bleedBaseDurations = {
+    [SPELLS.RAKE.id]: RAKE_BASE_DURATION,
+    [SPELLS.RIP.id]: RIP_BASE_DURATION,
+    [SPELLS.THRASH_FERAL.id]: THRASH_FERAL_BASE_DURATION,
+  }
+
+  earlyCastsOfTigersFury = 0;
+
+  // e.g. activeBleedsExpire[targetString][SPELLS.RAKE.id] = timestamp
+  activeBleedsExpire = {};
+  hasPredator = null;
+  hasJaggedWounds = null;
+
+  // timestamp of most recent possible kill event
+  possibleRecentKill = null;
+
+  // BFA: replace with constructor (and use this.selectedCombatant)
+  on_initialized() {
+    if (super.on_initialized) {
+      super.on_initialized();
+    }
+    this.hasPredator = this.combatants.selected.hasTalent(SPELLS.PREDATOR_TALENT.id);
+    this.hasJaggedWounds = this.combatants.selected.hasTalent(SPELLS.JAGGED_WOUNDS_TALENT.id);
+  }
+
+  on_byPlayer_applydebuff(event) {
+    if (super.on_byPlayer_applydebuff) {
+      super.on_byPlayer_applydebuff(event);
+    }
+    const spellId = event.ability.guid;
+    if (!this.hasPredator || !this.isBleed(spellId)) {
+      return;
+    }
+    const target = encodeTargetString(event.targetID, event.targetInstance);
+    if (!this.activeBleedsExpire[target]) {
+      this.activeBleedsExpire[target] = {};
+    }
+    const duration = this.constructor.bleedBaseDurations[spellId] * (this.hasJaggedWounds ? JAGGED_WOUNDS_MODIFIER : 1);
+    this.activeBleedsExpire[target][spellId] = event.timestamp + duration;
+  }
+
+  on_byPlayer_refreshdebuff(event) {
+    if (super.on_byPlayer_refreshdebuff) {
+      super.on_byPlayer_refreshdebuff(event);
+    }
+    const spellId = event.ability.guid;
+    if (!this.hasPredator || !this.isBleed(spellId)) {
+      return;
+    }
+    const target = encodeTargetString(event.targetID, event.targetInstance);
+    if (!this.activeBleedsExpire[target]) {
+      this.activeBleedsExpire[target] = {};
+    }
+    // existingExpire may be null if combat log missed the original applydebuff
+    const existingExpire = this.activeBleedsExpire[target][spellId];
+    const remainingOnPrevious = Math.max(0, existingExpire ? (event.timestamp - existingExpire) : 0);
```

Comment:
I'm a little confused here, shouldn't it be the opposite? My understanding:
- when you cast a bleed on a target, it calculates the "expected" timestamp when it should fall off on its own (line 64)
- but when you refresh it, then the bleed must still be on the target, so `existingExpire` should be larger than `event.timestamp` here, no?

Example:
- casted Rake at timestamp 120,000
- `this.activeBleedsExpire[target][SPELLS.RAKE.id]` should be 120,000 + 15,000 = 135,000 (assuming no Jagged Wounds)
- but when I *refresh* the Rake, it must have happened before timestamp 135,000, otherwise it wouldn't be a refresh but a new application
- so if I refresh Rake at timestamp let's say 130,000, 5 seconds before it should expire, `event.timestamp - existingExpire` = -5,000, so this line should evaluate *always* to 0

Am I missing something or understood it incorrectly?

Reply:
Thank you for that catch! Duration remaining on existing should definitely be `existingExpire - event.timestamp`